apiserver/params/apierror.go

```diff
@@ -18,30 +20,9 @@ var MigrationInProgressError = errors.New(CodeMigrationInProgress)
 
 // Error is the type of error returned by any call to the state API.
 type Error struct {
-	Message string     `json:"message"`
-	Code    string     `json:"code"`
-	Info    *ErrorInfo `json:"info,omitempty"`
-}
-
-// ErrorInfo holds additional information provided by an error.
-// Note that although these fields are compatible with the
-// same fields in httpbakery.ErrorInfo, the Juju API server does
-// not implement endpoints directly compatible with that protocol
-// because the error response format varies according to
-// the endpoint.
-type ErrorInfo struct {
-	// Macaroon may hold a macaroon that, when
-	// discharged, may allow access to the juju API.
-	// This field is associated with the ErrDischargeRequired
-	// error code.
-	Macaroon *macaroon.Macaroon `json:"macaroon,omitempty"`
-
-	// MacaroonPath holds the URL path to be associated
-	// with the macaroon. The macaroon is potentially
-	// valid for all URLs under the given path.
-	// If it is empty, the macaroon will be associated with
-	// the original URL from which the error was returned.
-	MacaroonPath string `json:"macaroon-path,omitempty"`
+	Message string                 `json:"message"`
```

Comment:
One of my initial concerns was how was this change going to impact the wire protocol.
From the look of this the existing expectations of non-juju clients (i.e. python-libjuju) is that the current implementation for deserializing the info portion will still match for the current implementation. The interesting part will be when we add an additional error info payload, like for the migrations. 
When we do this work, I'd like to see python-libjuju updated to handle the redirects. 

I have no concerns at this stage with how the changes are being sent.

Reply:
AFAICT from a quick look, python-libjuju [does not](https://github.com/juju/python-libjuju/blob/1ce8e0b5a650898b6df65cbcaaf18f532ee50456/juju/errors.py#L18-L22) currently process the `info` field. This is a bit confusing to me as `info` is only currently used for `DischargeRequiredError` and I would expect to see this in the python code (although there is a httpbakery.DischargeError in the [tests](https://github.com/juju/python-libjuju/blob/1ce8e0b5a650898b6df65cbcaaf18f532ee50456/tests/integration/test_macaroon_auth.py#L83)) 

Also, the [JujuRedirectException](https://github.com/juju/python-libjuju/blob/1ce8e0b5a650898b6df65cbcaaf18f532ee50456/juju/errors.py#L34) is currently built manually by making a second API call (works in JAAS but not implemented in upstream). This matches the implementation in juju cli. My next PR will allow this to happen as part of the login call so I can also open a PR against python-libjuju to ensure that it will work as expected.

Reply:
That would be awesome. There is an outstanding issue with python-libjuju around the discharge of macaroons.

https://github.com/juju/python-libjuju/pull/222